File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/page/admin/UserListPage.java
package net.dontdrinkandroot.gitki.wicket.page.admin;

import net.dontdrinkandroot.gitki.model.Role;
import net.dontdrinkandroot.gitki.model.User;
import net.dontdrinkandroot.gitki.wicket.column.UserActionColumn;
import net.dontdrinkandroot.gitki.wicket.component.button.UserAddButton;
import net.dontdrinkandroot.gitki.wicket.dataprovider.UserDataProvider;
import net.dontdrinkandroot.gitki.wicket.page.user.UserPage;
import net.dontdrinkandroot.gitki.wicket.security.Instantiate;
import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;
import net.dontdrinkandroot.wicket.bootstrap.css.TableStyle;
import org.apache.wicket.extensions.markup.html.repeater.data.table.DataTable;
import org.apache.wicket.extensions.markup.html.repeater.data.table.HeadersToolbar;
import org.apache.wicket.extensions.markup.html.repeater.data.table.IColumn;
import org.apache.wicket.extensions.markup.html.repeater.data.table.PropertyColumn;
import org.apache.wicket.extensions.markup.html.repeater.util.SortableDataProvider;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.request.mapper.parameter.PageParameters;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
@Instantiate(Role.ADMIN)
public class UserListPage extends UserPage<List<User>>
{
    public UserListPage(PageParameters parameters) {
        super(parameters);
    }

    @Override
    protected IModel<String> createTitleModel() {
        return Model.of("Users");
    }

    @Override
    protected void onInitialize() {
        super.onInitialize();

        this.add(new UserAddButton("addButton"));

        SortableDataProvider<User, String> dataProvider = new UserDataProvider();

        List<IColumn<User, String>> columns = new ArrayList<>();
        columns.add(new PropertyColumn<>(Model.of("Id"), "id", "id"));
        columns.add(new PropertyColumn<>(Model.of("Email"), "email", "email"));
        columns.add(new PropertyColumn<>(Model.of("First Name"), "firstName", "firstName"));
        columns.add(new PropertyColumn<>(Model.of("Last Name"), "lastName", "lastName"));
        columns.add(new PropertyColumn<>(Model.of("Role"), "role", "role"));
        columns.add(new UserActionColumn());

        DataTable<User, String> table = new DataTable<>("table", columns, dataProvider, 20);
        table.addTopToolbar(new HeadersToolbar<>(table, dataProvider));
        table.add(new CssClassAppender(BootstrapCssClass.TABLE));
        table.add(new CssClassAppender(TableStyle.STRIPED));
        table.add(new CssClassAppender(TableStyle.SMALL));
        this.add(table);
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/wicket/page/admin/UserListPage.kt
package net.dontdrinkandroot.gitki.wicket.page.admin

import net.dontdrinkandroot.gitki.model.Role
import net.dontdrinkandroot.gitki.model.User
import net.dontdrinkandroot.gitki.wicket.column.UserActionColumn
import net.dontdrinkandroot.gitki.wicket.component.button.UserAddButton
import net.dontdrinkandroot.gitki.wicket.dataprovider.UserDataProvider
import net.dontdrinkandroot.gitki.wicket.page.user.UserPage
import net.dontdrinkandroot.gitki.wicket.security.Instantiate
import net.dontdrinkandroot.wicket.bootstrap.behavior.TableBehavior
import net.dontdrinkandroot.wicket.bootstrap.css.TableStyle
import net.dontdrinkandroot.wicket.kmodel.kModel
import org.apache.wicket.extensions.markup.html.repeater.data.table.DataTable
import org.apache.wicket.extensions.markup.html.repeater.data.table.HeadersToolbar
import org.apache.wicket.extensions.markup.html.repeater.data.table.IColumn
import org.apache.wicket.extensions.markup.html.repeater.data.table.PropertyColumn
import org.apache.wicket.extensions.markup.html.repeater.util.SortableDataProvider
import org.apache.wicket.model.Model
import org.apache.wicket.model.StringResourceModel
import org.apache.wicket.request.mapper.parameter.PageParameters
import java.util.*

@Instantiate(Role.ADMIN)
class UserListPage(parameters: PageParameters = PageParameters()) : UserPage<List<User>>(parameters) {

    override fun createTitleModel() = kModel("Users")

    override fun onInitialize() {
        super.onInitialize()

        this.add(UserAddButton("addButton"))

        val dataProvider: SortableDataProvider<User, String> = UserDataProvider()
        val columns: MutableList<IColumn<User, String>> = ArrayList()
        columns.add(PropertyColumn(Model.of("Id"), "id", "id"))
        columns.add(PropertyColumn(StringResourceModel("gitki.user.email"), "email", "email"))
        columns.add(PropertyColumn(StringResourceModel("gitki.user.firstName"), "firstName", "firstName"))
        columns.add(PropertyColumn(StringResourceModel("gitki.user.lastName"), "lastName", "lastName"))
        columns.add(PropertyColumn(StringResourceModel("gitki.user.role"), "role", "role"))
        columns.add(UserActionColumn())
        val table = DataTable("table", columns, dataProvider, 20)
        table.addTopToolbar(HeadersToolbar(table, dataProvider))
        table.add(TableBehavior(null, TableStyle.STRIPED, TableStyle.SMALL))
        this.add(table)
    }
}